cli/cmd/inject.go

```diff
@@ -170,6 +170,11 @@ func (rt resourceTransformerInject) transform(bytes []byte) ([]byte, []inject.Re
 
 	reports := []inject.Report{*report}
 
+	if conf.IsService() && rt.values.Proxy.OpaquePorts != "" {
+		b, err := conf.AnnotateService(k8s.ProxyOpaquePortsAnnotation, rt.values.Proxy.OpaquePorts)
+		return b, reports, err
+	}
```

Comment:
This will need to be changed to deal with a few scenarios.

First, if we're injecting a service that does not have `rt.values.Proxy.OpaquePorts` set, then we don't branch into this and continue try and get a patch for a pod. This results in an error:

```
Error transforming resources:
add operation does not apply: doc is missing path: "/spec/template/spec/volumes/-": missing value
```

To deal with this, we should already plan to return when dealing with a service:

```suggestion
	if conf.IsService() {
		b := bytes
		if rt.values.Proxy.OpaquePorts != "" {
			b, err = conf.AnnotateService(k8s.ProxyOpaquePortsAnnotation, rt.values.Proxy.OpaquePorts)
		}
		return b, reports, err
```

Second, we can return a more helpful error message here so that we know that annotating the service failed. I'm thinking something like:

```
if err != nil {
	return nil, nil, fmt.Errorf("failed to annotate service %s: %s", report.Name, err)
}
```

Finally, this occurs before the `report.Injectable()` check below. If the service has `linker.io/inject: disabled`, we'll still annotate the service which we really shouldn't do. If inject is disabled on the service, then we should not mutate it at all.

This part is going to require changing what `func Injectable` checks. If we're dealing with a service object, we'll need to check the annotations on the object, where right now we only check annotations on a pod.

Reply:
Note that with #5766 merging, the `IsService` happens before the `Injectable` call ([code](https://github.com/linkerd/linkerd2/blob/edd3812f3050895fae1cc393d9d3365c405d6e21/cli/cmd/inject.go#L174-L176)).

This is okay for that change, because we never mutate the service YAML with the `inject` command. However for you change, this ordering does matter because we could be mutating the service. So we'll still want to move it below `Injectable` for this.